File: 프로그래머스/레벨2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    const bridge = Array(bridge_length).fill(0);
    let totalWeight = 0;
    let time = 0;

    for (let truck of truck_weights) {
        while (true) {
            time++;
            totalWeight -= bridge.shift();

            if (totalWeight + truck <= weight) {
                bridge.push(truck);
                totalWeight += truck;
                break;
            } else {
                bridge.push(0);
            }
        }
    }

    return time + bridge_length;
}

File: 프로그래머스/레벨2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> bridge = new LinkedList<>();
        int totalWeight = 0;        
        int time = 0;
        
        for (int i = 0; i < bridge_length; i++) {
            bridge.add(0);
        }        
        
        for (int truck : truck_weights) {
            while (true) {
                time++;
                totalWeight -= bridge.poll();

                if (totalWeight + truck <= weight) {
                    bridge.add(truck);
                    totalWeight += truck;
                    break;
                } else {
                    bridge.add(0);
                }
            }
        }

        return time + bridge_length;
    }
}